<a href="https://colab.research.google.com/github/khing-fm/Data_Science/blob/main/pandas_for_data_analysi.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Import pandas Iconvention: import as 'pd')
import pandas as pd
import numpy as np

print(f"pandas version: {pd.__version__}")
print(f"numpy version: {np.__version__}")

pandas version: 3.0.3
numpy version: 2.5.1


In [2]:
np.arange(0, 15, 2)

array([ 0,  2,  4,  6,  8, 10, 12, 14])

In [3]:
# Example: Creating a Series

# From a list
sales = pd.Series([100, 150, 200, 175, 225])
print("series from list:")
print(sales)
print(sales.index)
print(f"\nIndex: {sales.index.tolist()}")
print(f"Values: {sales.values.tolist()}")

series from list:
0    100
1    150
2    200
3    175
4    225
dtype: int64
RangeIndex(start=0, stop=5, step=1)

Index: [0, 1, 2, 3, 4]
Values: [100, 150, 200, 175, 225]


In [4]:
sales_mean = sales.mean()
sales_std = sales.std()

print(f"Mean Sales: {sales_mean}")
print(f"Standard Deviation of Sales: {sales_std}")

Mean Sales: 170.0
Standard Deviation of Sales: 48.088460154178364


### Sales Data Analysis Summary

*   **Mean Sales (170.0)**: Represents the average value of the sales entries.
*   **Standard Deviation (~48.09)**: Indicates the spread of the data; a higher value suggests that the sales figures vary significantly from the average.

In [5]:
pd.Series(np.array([10, 20, 30, 40, 50]), dtype="int16")

0    10
1    20
2    30
3    40
4    50
dtype: int16

### Creating a Series with NumPy and dtypes

This example demonstrates:
1.  **NumPy Integration**: Creating a Series directly from an `np.array`.
2.  **Memory Optimization**: Using `dtype="int16"` to store integers using less memory (16 bits) compared to the standard 64-bit default.

In [6]:
# Example: Series with custom index

sales = pd.Series(
    [100, 150, 200, 175, 225],
    index=["Jan", "Feb", "Mar", "Apr", "May"]
)
print("series with custom index:")
print(sales)
print(sales.index)
print(f"\nIndex: {sales.index.tolist()}")
print(f"Values: {sales.values.tolist()}")

# Acess by label
print(f"\nSales in May: {sales['May']}")
print(f"Jan to Mar: \n{sales['Jan':'Mar']}")

series with custom index:
Jan    100
Feb    150
Mar    200
Apr    175
May    225
dtype: int64
Index(['Jan', 'Feb', 'Mar', 'Apr', 'May'], dtype='str')

Index: ['Jan', 'Feb', 'Mar', 'Apr', 'May']
Values: [100, 150, 200, 175, 225]

Sales in May: 225
Jan to Mar: 
Jan    100
Feb    150
Mar    200
dtype: int64


### Custom Indexing and Label-Based Access

Pandas allows you to define a **custom index** (labels) for your data instead of just using default integers:

*   **Custom Labels**: In this example, we used month names (`Jan`, `Feb`, etc.) as the index.
*   **Access by Label**: You can retrieve specific values using the label name, like `sales['May']`.
*   **Slicing with Labels**: Unlike Python lists, when you slice with labels (`'Jan':'Mar'`), the **end point is included** in the result.

In [7]:
employee = {
    "id": "101",
    "name": "Alex Johnson",
    "email": "alex.johnson@example-pet-store.com",
    "is_active": True,
    "department": "HR"
}

In [8]:
import pandas as pd

pd.Series(employee)

id                                           101
name                                Alex Johnson
email         alex.johnson@example-pet-store.com
is_active                                   True
department                                    HR
dtype: object

In [9]:
staff_data = {
    "staff1":{
        "id": 101,
        "name": "John Doe",
        "email": "john.mclean@examplepetstore.com",
        "is_active": True,
        "department": "HR"
    },
    "staff2":{
        "id": 102,
        "name": "Jane Smith",
        "email": "john.catron@example-pet-store.com",
        "is_active": False,
        "department": "IT"
    },
    "staff3":{
        "id": 103,
        "name": "Bob Johnson",
        "email": "william.paterson@my-own-personal-domain.com",
        "is_active": True,
        "department": "Finance"
    },
    "staff4":{
        "id": 104,
        "name": "Alice Brown",
        "email": "john.mclean@examplepetstore.com",
        "is_active": True,
        "department": "Marketing"
    },
}

In [10]:
pd.DataFrame(staff_data)

,staff1,staff2,staff3,staff4
id,101,102,103,104
name,John Doe,Jane Smith,Bob Johnson,Alice Brown
email,john.mclean@examplepetstore.com,john.catron@example-pet-store.com,william.paterson@my-own-personal-domain.com,john.mclean@examplepetstore.com
is_active,True,False,True,True
department,HR,IT,Finance,Marketing


In [11]:
employee = {
    "name": ["John Doe", "Jane Smith", "Bob Johnson", "Alice Brown"],
    "id": [101, 102, 103, 104],
    "email": [
        "john.doe@example.com",
        "jane.smith@example.com",
        "bob.johnson@example.com",
        "alice.brown@example.com"
    ],
    "is_active": [True, False, True, True],
    "department": [["HR", "Finance"], ["IT"], ["Marketing"], ["HR", "Finance", "Marketing"]
    ]
}
# Dictionary of lists

In [12]:
pd.DataFrame(employee)

,name,id,email,is_active,department
0,John Doe,101,john.doe@example.com,True,"[HR, Finance]"
1,Jane Smith,102,jane.smith@example.com,False,[IT]
2,Bob Johnson,103,bob.johnson@example.com,True,[Marketing]
3,Alice Brown,104,alice.brown@example.com,True,"[HR, Finance, Marketing]"


In [13]:
df = pd.DataFrame(employee)
df.info()
df.shape
df.ndim
df.size
df.columns
df.index
df.values

<class 'pandas.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        4 non-null      str   
 1   id          4 non-null      int64 
 2   email       4 non-null      str   
 3   is_active   4 non-null      bool  
 4   department  4 non-null      object
dtypes: bool(1), int64(1), object(1), str(2)
memory usage: 264.0+ bytes


array([['John Doe', 101, 'john.doe@example.com', True,
        list(['HR', 'Finance'])],
       ['Jane Smith', 102, 'jane.smith@example.com', False, list(['IT'])],
       ['Bob Johnson', 103, 'bob.johnson@example.com', True,
        list(['Marketing'])],
       ['Alice Brown', 104, 'alice.brown@example.com', True,
        list(['HR', 'Finance', 'Marketing'])]], dtype=object)

In [14]:
df

,name,id,email,is_active,department
0,John Doe,101,john.doe@example.com,True,"[HR, Finance]"
1,Jane Smith,102,jane.smith@example.com,False,[IT]
2,Bob Johnson,103,bob.johnson@example.com,True,[Marketing]
3,Alice Brown,104,alice.brown@example.com,True,"[HR, Finance, Marketing]"


In [15]:
df [["name", "email", "department"]]

,name,email,department
0,John Doe,john.doe@example.com,"[HR, Finance]"
1,Jane Smith,jane.smith@example.com,[IT]
2,Bob Johnson,bob.johnson@example.com,[Marketing]
3,Alice Brown,alice.brown@example.com,"[HR, Finance, Marketing]"


In [16]:
pwd

'c:\\Users\\EHU_HEALTH_TECHNICAL\\Documents\\Khing_FM_Data_Science\\Data_Science'

In [17]:
# Reading CSV file
path = r"c:\\Users\\EHU_HEALTH_TECHNICAL\\Documents\\Khing_FM_Data_Science\\Data_Science\assets\ts-academy_datasets\\sales_data.csv"
# Read the sales data
sales_df = pd.read_csv(path)
sales_df

,transaction_id,date,product,category,quantity,unit_price,total_amount,region,sales_rep,customer_rating
0,TXN10001,2023-10-17,Wireless Mouse,Electronics,2,32.16,64.32,South,Bob Smith,5.0
1,TXN10002,2022-01-31,Wireless Mouse,Electronics,5,27.58,137.90,North,David Wilson,5.0
2,TXN10003,2022-10-12,Printer Paper,Office Supplies,6,21.29,127.74,West,Frank Miller,4.0
3,TXN10004,2023-01-25,Wireless Mouse,Electronics,5,33.12,165.60,South,Alice Johnson,5.0
4,TXN10005,2023-07-20,Monitor 27inch,Electronics,3,363.50,1090.50,East,Bob Smith,2.0
...,...,...,...,...,...,...,...,...,...,...
495,TXN10496,2022-06-01,Wireless Mouse,Electronics,5,26.23,131.15,West,Iris Chen,4.0
496,TXN10497,2022-05-14,Pen Set,Office Supplies,7,8.94,62.58,East,Jack Anderson,3.0
497,TXN10498,2022-03-15,Mechanical Keyboard,Electronics,5,127.26,636.30,North,Grace Lee,4.0
498,TXN10499,2022-01-25,Pen Set,Office Supplies,12,8.10,97.20,West,Henry Taylor,5.0


In [18]:
print(f"Shape: {sales_df.shape}")
print(f"Columns: {sales_df.columns.tolist()}")
print("\nFirst 5 rows:")
print(sales_df.head())

Shape: (500, 10)
Columns: ['transaction_id', 'date', 'product', 'category', 'quantity', 'unit_price', 'total_amount', 'region', 'sales_rep', 'customer_rating']

First 5 rows:
  transaction_id        date         product         category  quantity  \
0       TXN10001  2023-10-17  Wireless Mouse      Electronics         2   
1       TXN10002  2022-01-31  Wireless Mouse      Electronics         5   
2       TXN10003  2022-10-12   Printer Paper  Office Supplies         6   
3       TXN10004  2023-01-25  Wireless Mouse      Electronics         5   
4       TXN10005  2023-07-20  Monitor 27inch      Electronics         3   

   unit_price  total_amount region      sales_rep  customer_rating  
0       32.16         64.32  South      Bob Smith              5.0  
1       27.58        137.90  North   David Wilson              5.0  
2       21.29        127.74   West   Frank Miller              4.0  
3       33.12        165.60  South  Alice Johnson              5.0  
4      363.50       1090.50  

In [19]:
sales_df.describe()

,quantity,unit_price,total_amount,customer_rating
count,500.000000,490.000000,500.000000,482.000000
mean,7.050000,101.131184,390.324400,4.070539
std,5.887685,210.375358,754.739257,0.881316
min,1.000000,6.800000,7.580000,1.000000
25%,3.000000,12.270000,83.410000,4.000000
50%,5.000000,29.565000,142.700000,4.000000
75%,11.000000,52.797500,336.930000,5.000000
max,22.000000,1030.560000,7145.040000,5.000000


In [20]:
sales_df.dtypes

transaction_id         str
date                   str
product                str
category               str
quantity             int64
unit_price         float64
total_amount       float64
region                 str
sales_rep              str
customer_rating    float64
dtype: object

In [21]:
sales_df = pd.read_csv(
    path, parse_dates=["date"]
    )

In [22]:
sales_df.dtypes

transaction_id                str
date               datetime64[us]
product                       str
category                      str
quantity                    int64
unit_price                float64
total_amount              float64
region                        str
sales_rep                     str
customer_rating           float64
dtype: object

In [23]:
print(f"Date Column type: {sales_df['date'].dtype}")
print(f"\nDate range: {sales_df['date'].min()} to {sales_df['date'].max()}")

Date Column type: datetime64[us]

Date range: 2022-01-02 00:00:00 to 2024-01-01 00:00:00


In [24]:
print(f"\nNumber of unique products: {sales_df['product'].nunique()}")


Number of unique products: 15


In [25]:
# Reading JSON
path = r"c:\\Users\\EHU_HEALTH_TECHNICAL\\Documents\\Khing_FM_Data_Science\\Data_Science\assets\ts-academy_datasets\\products.json"

# Read the products data
products_json = pd.read_json(path)
print("JSON data:")
print(products_json)

JSON data:
                 catalog_version          last_updated currency  \
Electronics                  1.0  2024-01-15T10:30:00Z      USD   
Furniture                    1.0  2024-01-15T10:30:00Z      USD   
Office Supplies              1.0  2024-01-15T10:30:00Z      USD   

                                                        categories  
Electronics      {'tax_rate': 0.08, 'warranty_months': 12, 'pro...  
Furniture        {'tax_rate': 0.06, 'warranty_months': 24, 'pro...  
Office Supplies  {'tax_rate': 0.05, 'warranty_months': 0, 'prod...  


In [26]:
# Reading nested JSON (common with API responses   )

import json

# Load JSON file
with open(path, "r") as f:
    products_data = json.load(f)

# Extract electronics product into DataFrame
electronics = pd.DataFrame(products_data['categories']['Electronics']['products'])
# print("Electronics products:")
print(electronics)

        sku                        name    price  \
0  ELEC-001               Laptop Pro 15  1299.99   
1  ELEC-002              Wireless Mouse    29.99   
2  ELEC-003                  USB-C Dock    89.99   
3  ELEC-004         Mechanical Keyboard   149.99   
4  ELEC-005              4K Monitor 32"   449.99   
5  ELEC-006               Webcam HD Pro   129.99   
6  ELEC-007  Noise Canceling Headphones   279.99   
7  ELEC-008            Portable SSD 1TB   119.99   

                                           inventory  \
0  {'stock': 45, 'reorder_point': 10, 'supplier_i...   
1  {'stock': 350, 'reorder_point': 70, 'supplier_...   
2  {'stock': 120, 'reorder_point': 24, 'supplier_...   
3  {'stock': 78, 'reorder_point': 15, 'supplier_i...   
4  {'stock': 32, 'reorder_point': 10, 'supplier_i...   
5  {'stock': 95, 'reorder_point': 19, 'supplier_i...   
6  {'stock': 67, 'reorder_point': 13, 'supplier_i...   
7  {'stock': 145, 'reorder_point': 29, 'supplier_...   

                          